azure-iot-device/tests/iothub/transport/mqtt/test_mqtt_provider.py

```diff
@@ -8,164 +8,1120 @@
+        provider.on_mqtt_connected = callback
+
+        # Initiate reconnect
+        provider.reconnect(fake_password)
+
+        # Manually trigger Paho on_connect event_handler
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # Verify provider.on_mqtt_connected was called
+        assert callback.call_count == 1
+        assert callback.call_args == mocker.call()
+
+    @pytest.mark.it(
+        "Triggers on_mqtt_connected event handler callback upon externally-driven reconnect completion"
+    )
+    def test_calls_event_handler_callback_externally_driven(
+        self, mocker, mock_mqtt_client, provider
+    ):
+        callback = mocker.MagicMock()
+        provider.on_mqtt_connected = callback
+
+        # Manually trigger Paho on_connect event_handler
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # Verify provider.on_mqtt_connected was called
+        assert callback.call_count == 1
+        assert callback.call_args == mocker.call()
+
+    @pytest.mark.it(
+        "Skips on_mqtt_connected event handler callback if set to 'None' upon reconnect completion"
+    )
+    def test_skips_none_event_handler_callback(self, mocker, mock_mqtt_client, provider):
+        assert provider.on_mqtt_connected is None
+
+        provider.reconnect(fake_password)
+
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # No further asserts required - this is a test to show that it skips a callback.
+        # Not raising an exception == test passed
+
+    @pytest.mark.it("Recovers from exception in on_mqtt_connected event handler callback")
+    def test_event_handler_callback_raises_exception(self, mocker, mock_mqtt_client, provider):
+        event_cb = mocker.MagicMock(side_effect=DummyException)
+        provider.on_mqtt_connected = event_cb
+
+        provider.reconnect(fake_password)
+        mock_mqtt_client.on_connect(client=mock_mqtt_client, userdata=None, flags=None, rc=fake_rc)
+
+        # Callback was called, but exception did not propagate
+        assert event_cb.call_count == 1
+
+
+@pytest.mark.describe("MQTT Provider - Disconnect")
+class TestDisconnect(object):
+    @pytest.mark.it("Disconnects with Paho")
+    def test_calls_paho_disconnect(self, mocker, mock_mqtt_client, provider):
+        provider.disconnect()
+
+        assert mock_mqtt_client.disconnect.call_count == 1
+        assert mock_mqtt_client.disconnect.call_args == mocker.call()
+
+    @pytest.mark.it("Stops MQTT Network Loop")
+    def test_calls_loop_stop(self, mocker, mock_mqtt_client, provider):
+        provider.disconnect()
+
+        assert mock_mqtt_client.loop_stop.call_count == 1
+        assert mock_mqtt_client.loop_stop.call_args == mocker.call()
+
+    @pytest.mark.it(
+        "Triggers on_mqtt_disconnected event handler callback upon user-driven disconnect completion"
```

Comment:
>user-driven [](start = 67, length = 11)

same comment on wording.  I suggest searching the entire file for "user-driven" and "externally-driven" if you're going to change this.  I won't add any more comments on the matter.